google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,392 @@
+        )
+
+    def find_violations(self, res):
+        """Get a generator for violations
+
+        Args:
+            res (Resource): A class derived from Resource
+        Returns:
+            Generator: All violations of the resource breaking the rule.
+        """
+
+        if res.type == 'bucket':
+            return self.find_violations_in_bucket(res)
+        return iter(())
+
+    def bucket_max_retention_violation(self, bucket):
+        """Get a generator for violations especially for maximum retention
+           It only supports bucket for now, and will work on generalizing
+           in future PRs.
+
+        Args:
+            bucket (bucket): Find violation from the bucket
+        Yields:
+            RuleViolation: All max violations of the bucket breaking the rule.
+        """
+        if self.max_retention is None:
+            return
+
+        # There should be a condition which guarantees to delete data
+        bucket_lifecycle = bucket.get_lifecycle_rule()
+        if not bucket_lifecycle:
+            yield self.generate_bucket_violation(bucket)
+        else:
+            for lc_item in bucket_lifecycle:
+                if lc_item.get('action', {}).get('type') == 'Delete':
+                    conditions = lc_item.get('condition', {})
+                    age = conditions.get('age')
+                    if age is not None and len(conditions) == 1:
+                        # the config does not have conditions other than age
+                        if age <= self.max_retention:
+                            return
+            yield self.generate_bucket_violation(bucket)
+
+    def bucket_min_retention_violation(self, bucket):
+        """Get a generator for violations especially for minimum retention
+
+        Args:
+            bucket (bucket): Find violation from the bucket
+        Yields:
+            RuleViolation: All min violations of the bucket breaking the rule.
+        """
+        if self.min_retention is None:
+            return
+
+        bucket_lifecycle = bucket.get_lifecycle_rule()
+        for lc_item in bucket_lifecycle:
+            if (lc_item.get('action', {}).get('type') == 'Delete' and
+                    not bucket_conditions_guarantee_min(
+                        lc_item.get('condition', {}), self.min_retention)):
+                yield self.generate_bucket_violation(bucket)
+
+    def find_violations_in_bucket(self, bucket):
+        """Get a generator for violations
+        Args:
+            bucket (bucket): Find violation from the buckets
+        Returns:
+            Generator: All violations of the buckets breaking rules.
+        """
+
+        violation_max = self.bucket_max_retention_violation(bucket)
+        violation_min = self.bucket_min_retention_violation(bucket)
+        return itertools.chain(violation_max, violation_min)
+
+
+def bucket_conditions_guarantee_min(conditions, min_retention):
+    """check if other conditions can guarantee minimum retention
```

Comment:
Style nit: Capital letter at start of comment, and add a blank line between the first line (summary) and rest of the docstring.

(here and elsewhere)